File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_set=set()
        left=0
        max_length=0
        for right in range(len(s)):
            if s[right] not in char_set:
                char_set.add(s[right])
                max_length=max(max_length,right-left+1)
            else:
                while s[right]!=s[left]:
                    char_set.remove(s[left])
                    left+=1
                char_set.remove(s[left])
                left+=1
                char_set.add(s[right])
        return max_length

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character,Integer> map=new HashMap<>();
        int l=0,r=0;
        int n=s.length();
        int maxi=0;
        int len=0;
        while(r<n){
            char c=s.charAt(r);
            if(map.containsKey(c) && map.get(c)>=l){
                
                    l=map.get(c)+1;
                
            } 
            map.put(c,r);
            len=r-l+1;
            maxi=Math.max(maxi,len);
            r++;
        }
        return maxi;
    }
}